gui/petitions.lua

```diff
@@ -0,0 +1,301 @@
+        end
+    end
+
+    local petition = {}
+
+    if a.flags.petition_not_accepted then
+        sb[#sb+1] = {text = "This petition wasn't accepted yet!", pen = COLOR_YELLOW}
+        petition.status = 'PENDING'
+    elseif a.flags.convicted_accepted then
+        sb[#sb+1] = {text = "This petition was fulfilled!", pen = COLOR_GREEN}
+        petition.status = 'FULFILLED'
+    else
+        petition.status = 'ACCEPTED'
+    end
+    
+    petition.text = sb
+    
+    return petition
+end
+
+local getAgreements = function()
+    local list = {}
+
+    local ags = df.global.world.agreements.all
+    for i, a in ipairs(ags) do
+        for _, p in ipairs(a.parties) do
+            for _, e in ipairs(p.entity_ids) do
+                if e == we then
+                    list[#list+1] = getAgreementDetails(a)
+                end
+            end
+        end
+    end
+    
+    return list
+end
+
+local petitions = defclass(petitions, gui.FramedScreen)
+petitions.ATTRS = {
+    frame_style = gui.GREY_LINE_FRAME,
+    frame_title = 'Petitions',
+    frame_width = 21, -- is calculated in :refresh
+    min_frame_width = 21,
+    frame_height = 16,
+    frame_inset = 0,
+    focus_path = 'petitions',
+}
+
+function petitions:init(args)
+    self.list = args.list
+    -- self.fulfilled = true
+    self:addviews{
+        widgets.Label{
+            view_id = 'text',
+            frame_inset = 0,
+            text = '',
+        },
+    }
+    
+    self:refresh()
+end
+
+function petitions:refresh()
+    local lines = {}
+    -- list of petitions
+    for _, p in ipairs(self.list) do
+        if not self.fulfilled and p.status == 'FULFILLED' then goto continue end
+        -- each petition is a status and a text
+        for _, tok in ipairs(p.text) do
+            -- where text is a list of tokens
+            table.insert(lines, tok)
+        end
+        table.insert(lines, NEWLINE)
+    ::continue::
+    end
+    table.remove(lines, #lines) -- remove last NEWLINE
+
+    local label = self.subviews.text
+    label:setText(lines)
+    
+    -- changing text doesn't automatically change scroll position
+    if label.frame_body then
+        local last_visible_line = label.start_line_num + label.frame_body.height - 1
+        if last_visible_line > label:getTextHeight() then
+            label.start_line_num = math.max(label:getTextHeight() - label.frame_body.height + 1, 1)
+        end
+    end
+    
+    self.frame_width = math.max(label:getTextWidth(), self.min_frame_width)
+    self.frame_width = math.min(df.global.gps.dimx - 2, self.frame_width)
+    self:onResize(dfhack.screen.getWindowSize()) -- applies new frame_width
+end
+
+function petitions:onRenderFrame(painter, frame)
+    petitions.super.onRenderFrame(self, painter, frame)
+    
+    painter:seek(frame.x1+2, frame.y1 + frame.height-1):key_string('CUSTOM_F', "toggle fulfilled")
+    
+    local label = self.subviews.text
+    if label:getTextHeight() > label.frame_body.height then
```

Comment:
This looks like it should be added to the Label class. Would it still look good if rendered on the label widget instead of the frame boundary?

Reply:
I personally prefer the arrows and the text rendered on the frame. Second-best good-looking is arrows on an extra column on the very right.

I think both are possible to implement, although the second option is probably much easier.

Here some images for comparison:

on the border:
![image](https://user-images.githubusercontent.com/348150/162567482-d1fd03ec-dabc-4bc8-9b4a-bb876770d39c.png)

on an extra right-most column:
![image](https://user-images.githubusercontent.com/348150/162569050-f0f7b063-08ae-4f84-a254-0046971dd9de.png)

arrows on the label widget, no inset (2nd image illustrates a problem with that, notice how arrow hides a character):
![image](https://user-images.githubusercontent.com/348150/162568753-35180d12-dc66-4e48-9027-0d8f53359d17.png)
![image](https://user-images.githubusercontent.com/348150/162568765-36f3f257-07b8-43f3-9626-7a51a9223aa5.png)

label-widget with `frame_inset = 1`
![image](https://user-images.githubusercontent.com/348150/162568095-1b842930-3892-4334-95e2-1f7f7a6e3de2.png)

Reply:
Thank you for visualizing the options!

Visual-wise, I think the arrows are easiest to see when they're inside the frame. We have a long-standing standard of putting hotkey hints on the bottom frame (and they look fine there too), so I vote for arrows inside frame, hotkey text on bottom frame.

Architecture-wise, the arrow logic should be part of the Label widget. Labels are used all over the place, and they don't always have a frame around them, but they're always scrollable. We could potentially have the logic in the Label widget but render it somewhere else. That might cause confusion, though.

Behavior-wise, the arrows should only appear if the content height is larger than the Label body area height, otherwise existing, static labels would get unneeded arrows. If they appear, maybe they should force at least a right margin of 1 to avoid the character hiding problem you highlighted. If the widget already has a frame_inset>=1 (or frame_inset.r>=1) then this is a noop, otherwise, just the right inset can be increased by 1. The arrows should probably align vertically with the top and bottom frame inset, and be in the column just to the right of the end of the text area, but I guess check what looks best if frame inset is like 5 or something.

Thank you for doing this! This will be an awesome addition to the Label widget!

Reply:
I've created a WIP pull request: https://github.com/DFHack/dfhack/pull/2101 (there is a problem I'm not quite sure what to do about)

Reply:
https://github.com/DFHack/dfhack/pull/2101 is now merged. Thanks!